pkg/provision/primitives/vmutils.go

```diff
@@ -0,0 +1,264 @@
+	}
+	// load IP
+	ip, ok := workloadDefinition.(*workloads.PublicIP)
+	if !ok {
+		return net.IPNet{}, nil, errors.Wrap(err, "could not decode ip reservation")
+	}
+	identity := stubs.NewIdentityManagerStub(p.zbus)
+	self := identity.NodeID().Identity()
+	selfDescription, err := explorerClient.Directory.NodeGet(self, false)
+	if err != nil {
+		return net.IPNet{}, nil, errors.Wrap(err, "could not get our own node description")
+	}
+	farm, err := explorerClient.Directory.FarmGet(schema.ID(selfDescription.FarmId))
+	if err != nil {
+		return net.IPNet{}, nil, errors.Wrap(err, "could not get our own farm")
+	}
+
+	var pubGw schema.IP
+	for _, ips := range farm.IPAddresses {
+		if ips.ReservationID == rid {
+			pubGw = ips.Gateway
+			break
+		}
+	}
+	if pubGw.IP == nil {
+		return net.IPNet{}, nil, errors.New("unable to identify public ip gateway")
+	}
+
+	return ip.IPaddress.IPNet, pubGw.IP, nil
+}
+
+// returns the vCpu's, memory, disksize for a vm size
+// memory and disk size is expressed in MiB
+func vmSize(vm VMWithCustomSize) (cpu uint8, memory uint64, storage uint64, err error) {
+	switch vm.GetSize() {
+	case -1:
+		customSize := vm.GetCustomSize()
+		return uint8(customSize.CRU),
+			uint64(customSize.MRU * 1024),
+			uint64(customSize.SRU * 1024),
+			nil
+	case 1:
+		// 1 vCpu, 2 GiB RAM, 50 GB disk
+		return 1, 2 * 1024, 50 * 1024, nil
+	case 2:
+		// 2 vCpu, 4 GiB RAM, 100 GB disk
+		return 2, 4 * 1024, 100 * 1024, nil
+	case 3:
+		// 2 vCpu, 8 GiB RAM, 25 GB disk
+		return 2, 8 * 1024, 25 * 1024, nil
+	case 4:
+		// 2 vCpu, 8 GiB RAM, 50 GB disk
+		return 2, 8 * 1024, 50 * 1024, nil
+	case 5:
+		// 2 vCpu, 8 GiB RAM, 200 GB disk
+		return 2, 8 * 1024, 200 * 1024, nil
+	case 6:
+		// 4 vCpu, 16 GiB RAM, 50 GB disk
+		return 4, 16 * 1024, 50 * 1024, nil
+	case 7:
+		// 4 vCpu, 16 GiB RAM, 100 GB disk
+		return 4, 16 * 1024, 100 * 1024, nil
+	case 8:
+		// 4 vCpu, 16 GiB RAM, 400 GB disk
+		return 4, 16 * 1024, 400 * 1024, nil
+	case 9:
+		// 8 vCpu, 32 GiB RAM, 100 GB disk
+		return 8, 32 * 1024, 100 * 1024, nil
+	case 10:
+		// 8 vCpu, 32 GiB RAM, 200 GB disk
+		return 8, 32 * 1024, 200 * 1024, nil
+	case 11:
+		// 8 vCpu, 32 GiB RAM, 800 GB disk
+		return 8, 32 * 1024, 800 * 1024, nil
+	case 12:
+		// 1 vCpu, 64 GiB RAM, 200 GB disk
+		return 1, 64 * 1024, 200 * 1024, nil
+	case 13:
+		// 1 mvCpu, 64 GiB RAM, 400 GB disk
+		return 1, 64 * 1024, 400 * 1024, nil
+	case 14:
+		//1 vCpu, 64 GiB RAM, 800 GB disk
+		return 1, 64 * 1024, 800 * 1024, nil
+	case 15:
+		//1 vCpu, 2 GiB RAM, 25 GB disk
+		return 1, 2 * 1024, 25 * 1024, nil
+	case 16:
+		//2 vCpu, 4 GiB RAM, 50 GB disk
+		return 2, 4 * 1024, 50 * 1024, nil
+	case 17:
+		//4 vCpu, 8 GiB RAM, 50 GB disk
+		return 4, 8 * 1024, 50 * 1024, nil
+	case 18:
+		//1 vCpu, 1 GiB RAM, 25 GB disk
+		return 1, 1 * 1024, 25 * 1024, nil
+	}
+
+	return 0, 0, 0, fmt.Errorf("unsupported vm size %d, only size 1 to 18 are supported", vm.GetSize())
```

Comment:
-1, and 1 to 18 are supported

Reply:
-1 is a flag that a custom size is provided not a predefined size. So I feel it might confuse the user (that he can provide -1 and everything will work as if he provided any other number). Do you think it should be mentioned in the message?

Reply:
I'm actually not sure, it's a valid value for the size, but probably @muhamadazmy can decide on this

Reply:
The custom size is filled up by the explorer (user can't provide that) so if a user asked for size `-1` the explorer will fill the custom size in the reservation to max that this node can support